docs/lang/articles/misc/ndarray_android.md

```diff
@@ -0,0 +1,449 @@
+
+- **Kernel-specific** `args_buff`
+
+  The `initialize` kernel is assigned an `args_buffer` of `128` Bytes (line 21). Note that the size of `args_buffer` is dependent on the number of Ndarrays (`pos` and `vel`) that the kernel takes, (see `arg_count` in line 19). The `initialize` kernel, or each kernel more precisely, has a dedicated `args_buffer` for storing scalar arguments specified in `scalar_args` (line 27) and Ndarray shape information in accordance with what `array_args` (line 28-45) specifies.
+
+  Ndarrays' shape information is organized by their argument index in the `array_args` JSON array: `0` (line 29) corresponds to the `pos` Ndarray, and `1` (line 37) corresponds to the `vel` Ndarray. The argument index is determined by the sequence by which you pass in the Ndarrays when calling `add_kernel()`. See line 53 in the Python script.
+
+  The `pos` Ndarray's shape information in `args_buffer` has an offset of  `64` Bytes in `args_buffer` (line 64). According to line 35 and line 43, the `pos` Ndarray's shape information occupies 96 - 64 = 32 Bytes in `args_buffer`.
+
+  :::tip ATTENTION
+  The JSON file only specifies the dimension of the corresponding Ndarray (line 30, 38), allowing you to dynamically update an Ndarray's shape in your runtime program.
+  :::
+
+- **Kernel-specific binding index**
+
+  `used.arr_arg_to_bind_idx` (line 46) maps the SSBO of each Ndarray in the kernel to a "more global" binding index for the compute shaders. For example, `"1": 5` (line 48) binds the `vel` Ndarray to the binding index `5`.
+
+```json
+{
+  "aot_data": {
+    "kernels": {
+      "initialize": {
+        "tasks": [
+          {
+            "name": "initialize_c54_00",
+            "src": "nbody_aot/initialize_c54_00.glsl",
+            "workgroup_size": 1,
+            "num_groups": 1
+          },
+          {
+            "name": "initialize_c54_01",
+            "src": "nbody_aot/initialize_c54_01.glsl",
+            "workgroup_size": 128,
+            "num_groups": 256
+          }
+        ],
+        "arg_count": 2,
+        "ret_count": 0,
+        "args_buf_size": 128,
+        "ret_buf_size": 0,
+        "ext_arr_access": {
+          "0": 2,
+          "1": 3
+        },
+        "scalar_args": {},
+        "arr_args": {
+          "0": {
+            "field_dim": 1,
+            "is_scalar": false,
+            "element_shape": [
+              2
+            ],
+            "shape_offset_in_bytes_in_args_buf": 64
+          },
+          "1": {
+            "field_dim": 1,
+            "is_scalar": false,
+            "element_shape": [
+              2
+            ],
+            "shape_offset_in_bytes_in_args_buf": 96
+          }
+        },
+        "used.arr_arg_to_bind_idx": {
+          "0": 4,
+          "1": 5
+        }
+      },
+      "compute_force": {
```

Comment:
If we are using `init` kernel as example, shall we just omit most lines for `compute_force` to keep this article short?

Reply:
+1, I think the complete JSON file can be checked in to the repo itself